app/scripts/modules/core/src/pipeline/config/stages/googleCloudBuild/GoogleCloudBuildStageConfig.tsx

```diff
@@ -1,87 +1,36 @@
 
-import { GoogleCloudBuildStageForm, buildDefinitionSources, triggerType } from './GoogleCloudBuildStageForm';
+import { GoogleCloudBuildStageForm } from './GoogleCloudBuildStageForm';
 import { validate } from './googleCloudBuildValidators';
-
-interface IGoogleCloudBuildStageConfigState {
-  googleCloudBuildAccounts: string[];
-  gcbTriggers: IGcbTrigger[];
-}
-
-export class GoogleCloudBuildStageConfig extends React.Component<IStageConfigProps, IGoogleCloudBuildStageConfigState> {
-  private stage: IStage;
-  private destroy$ = new Subject();
-
-  public constructor(props: IStageConfigProps) {
-    super(props);
-    this.state = {
-      googleCloudBuildAccounts: [],
-      gcbTriggers: [],
-    };
-    const { stage: initialStageConfig } = props;
-    const stage = cloneDeep(initialStageConfig);
-    if (!stage.application) {
-      stage.application = props.application.name;
-    }
-    if (!stage.buildDefinitionSource) {
-      stage.buildDefinitionSource = buildDefinitionSources.TEXT;
-    }
-    if (!stage.triggerType) {
-      stage.triggerType = triggerType.BRANCH;
-    }
-    // Intentionally initializing the stage config only once in the constructor
-    // The stage config is then completely owned within FormikStageConfig's Formik state
-    this.stage = stage;
-  }
-
-  public componentDidMount = (): void => {
-    this.fetchGoogleCloudBuildAccounts();
-    if (this.stage.account) {
-      this.fetchGcbTriggers(this.stage.account);
-    }
-  };
-
-  private fetchGoogleCloudBuildAccounts = (): void => {
-    Observable.fromPromise(IgorService.getGcbAccounts())
-      .takeUntil(this.destroy$)
-      .subscribe((googleCloudBuildAccounts: string[]) => {
-        this.setState({ googleCloudBuildAccounts });
-      });
-  };
-
-  public componentWillUnmount(): void {
-    this.destroy$.next();
-  }
-
-  private fetchGcbTriggers: (account: string) => void = account => {
-    Observable.fromPromise(IgorService.getGcbTriggers(account))
-      .takeUntil(this.destroy$)
-      .subscribe((gcbTriggers: IGcbTrigger[]) => {
-        this.setState({ gcbTriggers });
-      });
-  };
-
-  public render() {
-    return (
-      <FormikStageConfig
-        {...this.props}
-        stage={this.stage}
-        onChange={this.props.updateStage}
-        validate={validate}
-        render={props => (
-          <GoogleCloudBuildStageForm
-            {...props}
-            googleCloudBuildAccounts={this.state.googleCloudBuildAccounts}
-            gcbTriggers={this.state.gcbTriggers}
-            updatePipeline={this.props.updatePipeline}
-            fetchGcbTriggers={this.fetchGcbTriggers}
-          />
-        )}
-      />
-    );
-  }
+import { BuildDefinitionSource, IGoogleCloudBuildStage } from './IGoogleCloudBuildStage';
+
+export function GoogleCloudBuildStageConfig({
+  application,
+  pipeline,
+  stage,
+  updatePipeline,
+  updateStage,
+}: IStageConfigProps) {
+  const stageWithDefaults: IGoogleCloudBuildStage = React.useMemo(() => {
```

Comment:
Any thoughts on a better way to initialize a stage config with defaults?

Reply:
This seems to work. An alternative, native way for doing this could be `{ application: application.name, buildDefinitionSource: buildDefinitionSource.TEXT, ...cloneDeep(stage)  }`

Reply:
:+1: :zap: updated, thanks! Was mostly curious if anyone had any thoughts on where the defaulting should happen (in the parent component that renders the `FormikStageConfig` vs. in the actual stage config component).